Introducing ReadOrigin. Fix #477

JimBobSquarePants opened the pull request:
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
This PR introduces new overloads for any of our methods that accept a stream. Fixing #477

These overloads contain a new parameter `ReadOrigin` which allows the calling code to instruct us to read from either the beginning or the current position within the stream, defaulting to the beginning.

This ensures that developers are not surprised, nor have to reset the stream position in a common use case.
 
<!-- Thanks for contributing to ImageSharp! -->

tocsoft commented:
It feels like it would make more sense to put the `ReadOrigin` onto the `Configuration` object instead of just passing it into all the places we use streams, thus it can be globally opted into.

antonfirsov commented:
@JimBobSquarePants I think there is a misunderstanding. Isn't the point of #477 that we are failing to load from a non-seekable stream?

In that case this is only a "fix" for the workaround that asks the user to manually load the image into a memory stream. We are just rewinding the stream to the beginning, so the user shouldn't.

If my understanding is correct, I think we should not do this, this is quite unconventional. We should investigate the issue with non-seekable streams instead.

tocsoft commented:
if the stream is non-seekable we are already copying it internally into a memory stream then reading it from the beginning... re the original issue is we must be detecting it as seekable thus not doing the internal copy and thus failing to reset it between attempts to detect image format.

JimBobSquarePants commented:
@antonfirsov There's actually two issues embedded in one there, I missed the second issue that simply passing the stream doesn't work.

@tocsoft I had a look through the AWS code. `MD5Stream` is both non-readable and non-seekable. We're throwing because it's non-readable and `CopyTo()` relies on the stream being readable. 

However, the OP managed to wrap the stream in a `MemoryStream` (I'm assuming through the constructor but I don't want to have to download the SDK to check) so I've asked him how. We should use that approach rather than our existing one if this is the case.

Regarding using `Configuration`, I designed it like `Stream.Seek(SeekOrigin)` which the localized context works well with. I guess I could move it there though, it might actually make the code a little cleaner.

dlemstra commented on src/ImageSharp/Configuration.cs:
```diff
@@ -83,6 +83,11 @@ public Configuration(params IConfigurationModule[] configurationModules)
+        /// <summary>
+        /// Gets or sets the position in a stream to use for reading.
+        /// </summary>
+        public ReadOrigin ReadOrigin { get; set; } = ReadOrigin.Begin;
```
I don't think we should make this the default. We should not seek within the stream unless the user instructs us to do so.

JimBobSquarePants replied:
Based on my normal expectations I would be inclined to think the same. However, I think many other devs would expect it to JustWork™

dlemstra replied:
I personally think this is a bad decision but you are the "Grand High Eternal Dictator" so you have the final say. I do think we need an option for this and it is a great idea to add this but this should not be the default.

JimBobSquarePants replied:
Haha! We'll see what the other chaps think. Off to bed now so will come back to it in the morning.

antonfirsov replied:
I'm not convinced at all that we need to do this, specially not by default!
See my comment on #477.
Or am I missing something? 😕

dlemstra commented on src/ImageSharp/Image/Image.FromStream.cs:
```diff
@@ -246,9 +241,19 @@ private static T WithSeekableStream<T>(Stream stream, Func<Stream, T> action)
                 return action(stream);
             }
 
+            if (config.ReadOrigin == ReadOrigin.Current)
```
Shouldn't this be `config.ReadOrigin != ReadOrigin.Current` or `config.ReadOrigin == ReadOrigin.Begin`?

JimBobSquarePants replied:
If a stream is non-seekable then it's always gonna end up at 0 since we copy it to a `MemoryStream`. I thought best to throw rather than silently ignore.

dlemstra replied:
Maybe I don't understand what this method should do but shouldn't it copy the data from the stream to a memory stream to make it seekable? The ReadOrigin is an option for the input stream or also for the stream that we create?

tocsoft replied:
I feel we should be using this setting as more of a hint than a hard rule. Basically we use it as permission to reset streams when they are seekable and when they are not then we use our current logic of copy and read from there.

JimBobSquarePants commented:
I'm putting this on hold for the moment. Something funky is going on.
https://github.com/SixLabors/ImageSharp/issues/477#issuecomment-369424708

JimBobSquarePants commented:
Ok... So it turns out #477 was user error. 

I think there's still merit in adding some granular control to our stream seeking behavior so with that in mind I've changed the default to `ReadOrigin.Current` and removed the extra exception. 

Default behaviour is still the same as prior to the PR, we just have extra options now.
